lib/dal/src/attribute/value/dependent_value_graph.rs

```diff
@@ -158,6 +165,82 @@ impl DependentValueGraph {
         Ok(values)
     }
 
+    async fn process_subscription_dependencies(
+        &mut self,
+        ctx: &DalContext,
+        component_root_attribute_value_id: AttributeValueId,
+        workspace_snapshot: &WorkspaceSnapshotSelector,
+    ) -> AttributeValueResult<Vec<WorkQueueValue>> {
+        let current_component_id = self
+            .component_id_for_av(ctx, component_root_attribute_value_id)
+            .await?;
+
+        let mut new_work_queue_values = vec![];
+
+        for (edge, subscriber_apa_id, _subscribed_to_av_id) in workspace_snapshot
+            .edges_directed(component_root_attribute_value_id, Direction::Incoming)
+            .await?
+        {
+            if let EdgeWeightKind::ValueSubscription(path) = edge.kind {
+                let subscription = ValueSubscription {
+                    attribute_value_id: component_root_attribute_value_id,
+                    path,
+                };
+                if let Some(resolved_av_id) = subscription.resolve(ctx).await? {
+                    if let Some(subscriber_ap_id) = workspace_snapshot
+                        .source_opt(subscriber_apa_id, EdgeWeightKind::PrototypeArgument)
+                        .await?
+                    {
+                        for subscriber_av_ulid in workspace_snapshot
+                            .incoming_sources_for_edge_weight_kind(
+                                subscriber_ap_id,
+                                EdgeWeightKindDiscriminants::Prototype,
+                            )
+                            .await?
+                        {
+                            let subscriber_av_id = subscriber_av_ulid.into();
+                            let subscriber_component_id =
+                                self.component_id_for_av(ctx, subscriber_av_id).await?;
+
+                            if subscriber_component_id != current_component_id {
+                                if !Component::should_data_flow_between_components(
+                                    ctx,
+                                    subscriber_component_id,
+                                    current_component_id,
+                                )
+                                .await?
+                                {
+                                    continue;
+                                }
+
+                                // For an inter-component subscription, the subscriber should depend on the other
+                                // component's *root* attribute value, so that the entire other component's data is
+                                // processed by the dvu and available to the subscriber before we try to run the
+                                // subscriber's function.
+                                self.value_depends_on(
+                                    subscriber_av_id,
+                                    component_root_attribute_value_id,
+                                );
```

Comment:
Why is depending on the resolved subscribed to AV not sufficient? The subscription shouldn't be using any data that isn't the resolved subscribed to AV or one of its descendants.

Reply:
On testing again, the new test fails if we only make the subscribe_av_id depend on the resolved_av_id. I believe this is because of the way the subscriptions resolve values when they're executing, but I'd have to investigate it further.

Reply:
Yeah, having to depend on the root feels like papering over a bug elsewhere that should be fixed instead.

Reply:
Using the "controlling avs" for each seems to work

Reply:
That sounds like the right thing to do, since that'll be the function that generates the value (if it's not from the subscribed-to AV's prototype). 👍🏻